pulsar-broker/src/main/java/org/apache/pulsar/compaction/PulsarTopicCompactionService.java

```diff
@@ -106,6 +111,78 @@ public CompletableFuture<Position> getLastCompactedPosition() {
         return CompletableFuture.completedFuture(compactedTopic.getCompactionHorizon().orElse(null));
     }
 
+    @Override
+    public CompletableFuture<Entry> findEntryByPublishTime(long publishTime) {
+        final Predicate<Entry> predicate = entry -> {
+            try {
+                return Commands.parseMessageMetadata(entry.getDataBuffer()).getPublishTime() >= publishTime;
+            } finally {
+                entry.release();
+            }
+        };
+        return findFirstMatchEntry(predicate);
+    }
+
+    @Override
+    public CompletableFuture<Entry> findEntryByEntryIndex(long entryIndex) {
+        final Predicate<Entry> predicate = entry -> {
+            try {
+                BrokerEntryMetadata brokerEntryMetadata =
+                        Commands.parseBrokerEntryMetadataIfExist(entry.getDataBuffer());
+                if (brokerEntryMetadata == null || !brokerEntryMetadata.hasIndex()) {
+                    return false;
+                }
+                return brokerEntryMetadata.getIndex() >= entryIndex;
+            } finally {
+                entry.release();
+            }
+        };
+        return findFirstMatchEntry(predicate);
+    }
+
+    private CompletableFuture<Entry> findFirstMatchEntry(final Predicate<Entry> predicate) {
+        var compactedTopicContextFuture = compactedTopic.getCompactedTopicContextFuture();
+
+        if (compactedTopicContextFuture == null) {
+            return CompletableFuture.completedFuture(null);
+        }
+        return compactedTopicContextFuture.thenCompose(compactedTopicContext -> {
+            LedgerHandle lh = compactedTopicContext.getLedger();
+            CompletableFuture<Long> promise = new CompletableFuture<>();
+            findFirstMatchIndexLoop(predicate, 0L, lh.getLastAddConfirmed(), promise, null, lh);
+            return promise.thenCompose(index -> {
+                if (index == null) {
+                    return CompletableFuture.completedFuture(null);
+                }
+                return readEntries(lh, index, index).thenApply(entries -> entries.get(0));
+            });
+        });
+    }
+
+    private static void findFirstMatchIndexLoop(final Predicate<Entry> predicate,
+                                           final long start, final long end,
+                                           final CompletableFuture<Long> promise,
+                                           final Long lastMatchIndex,
+                                           final LedgerHandle lh) {
+        if (start > end) {
+            promise.complete(lastMatchIndex);
+            return;
+        }
+
+        long mid = (start + end) / 2;
+        CompactedTopicImpl.readEntries(lh, mid, mid).thenAccept(entries -> {
```

Comment:
You already imported `import static org.apache.pulsar.compaction.CompactedTopicImpl.readEntries;`. You can get rid of `CompactedTopicImpl` prefix.

Reply:
Ok.